Coder Desktop/Coder Desktop/VPNMenuState.swift

```
@@ -0,0 +1,116 @@
+import Foundation
+import SwiftUI
+import VPNLib
+
+struct Agent: Identifiable, Equatable, Comparable {
+    let id: UUID
+    let name: String
+    let status: AgentStatus
+    let copyableDNS: String
+    let wsName: String
+    let wsID: UUID
+
+    // Agents are sorted by status, and then by name
+    static func < (lhs: Agent, rhs: Agent) -> Bool {
+        if lhs.status != rhs.status {
+            return lhs.status < rhs.status
+        }
+        return lhs.wsName.localizedCompare(rhs.wsName) == .orderedAscending
+    }
+}
+
+enum AgentStatus: Int, Equatable, Comparable {
+    case okay = 0
+    case warn = 1
+    case error = 2
+    case off = 3
+
+    public var color: Color {
+        switch self {
+        case .okay: .green
+        case .warn: .yellow
+        case .error: .red
+        case .off: .gray
+        }
+    }
+
+    static func < (lhs: AgentStatus, rhs: AgentStatus) -> Bool {
+        lhs.rawValue < rhs.rawValue
+    }
+}
+
+struct Workspace: Identifiable, Equatable, Comparable {
+    let id: UUID
+    let name: String
+    var agents: [UUID]
+
+    static func < (lhs: Workspace, rhs: Workspace) -> Bool {
+        lhs.name.localizedCompare(rhs.name) == .orderedAscending
+    }
+}
+
+struct VPNMenuState {
+    var agents: [UUID: Agent] = [:]
+    var workspaces: [UUID: Workspace] = [:]
+
+    mutating func upsertAgent(_ agent: Vpn_Agent) {
+        guard let id = UUID(uuidData: agent.id) else { return }
+        guard let wsID = UUID(uuidData: agent.workspaceID) else { return }
+        // An existing agent with the same name, belonging to the same workspace
+        // is from a previous workspace build, and should be removed.
+        agents.filter { $0.value.name == agent.name && $0.value.wsID == wsID }
+            .forEach { agents[$0.key] = nil }
+        workspaces[wsID]?.agents.append(id)
+        let wsName = workspaces[wsID]?.name ?? "Unknown Workspace"
```

Under what circumstances might the `name` and `copyableDNS` be empty?

Wouldn't that make the menu entry useless as well?

---

We always upsert workspaces before agents, so `name` would only ever be empty if the server for whatever reason gave us an agent upsert before a corresponding workspace upsert. It shouldn't happen, but I also didn't want to swallow that case and not provide any indication it happened - though this might not be the best way?

---

`copyableDNS` also shouldn't ever be nil, but we're not guaranteed the fqdn list on the protobuf is non-empty. The list is always populated with at least one FDQN within `tailnet`, but again, if something does go wrong I don't want to it to just be ignored.

---

> I also didn't want to swallow that case and not provide any indication it happened
> if something does go wrong I don't want to it to just be ignored.

Not wanting to hide errors is understandable. I'm trying to understand how "Unknown Workspace" and "UNKNOWN" will help clarify whether this information was received out of order or is missing certain entries.

An indicator in the menu dropdown (similar to Tailscale when DNS isn't functioning correctly) would present and clarify the issue more clearly and in greater detail about what went wrong. Potentially even opening a popup window with more details when clicking on it.

Currently, we would add an unusable entry to the list, potentially even multiple entries at once, without communicating the problem to the user.

---

Yeah, that's a good point - I'll add a way to plumb through non-fatal error messages into the VPNState view so an alert can be shown.

---

In that case, I'd advocate that the list include only correct entries, and any unexpected items can be placed in the error view/section.
If data arrives later, we can "restore" the entry back into the list as it will be correct and complete by that point.